### PR TITLE
Isochrone : return several isochrones in response

### PR DESCRIPTION
**Pull request by ZoeBrunet**

### CODE REVIEW

File: source/jormungandr/jormungandr/scenarios/simple.py [hunk 1]
```diff
@@ -228,19 +227,32 @@ def graphical_isochrones(self, request, instance):
...
 
-        journey_req.max_duration = request["max_duration"]
         journey_req.max_transfers = request["max_transfers"]
         self.origin_modes = request["origin_mode"]
         self.destination_modes = request["destination_mode"]
         if "forbidden_uris[]" in request and request["forbidden_uris[]"]:
             for forbidden_uri in request["forbidden_uris[]"]:
                 req.journeys.forbidden_uris.append(forbidden_uri)
+        req.isochrone.min_duration = request["min_duration"]
+        if "max_duration" in request and request["max_duration"]:
```
Comment (xlqian) [inline source/jormungandr/jormungandr/scenarios/simple.py]: `request.get('max_duration')` ?

File: source/jormungandr/jormungandr/scenarios/simple.py [hunk 2]
```diff
@@ -228,19 +227,32 @@ def graphical_isochrones(self, request, instance):
...
+            journey_req.max_duration = request["max_duration"]
+        else:
+            journey_req.max_duration = max(request["duration[]"], key=int)
+        if "duration[]" in request and request["duration[]"]:
```
Comment (xlqian) [inline source/jormungandr/jormungandr/scenarios/simple.py]: same as above

File: source/jormungandr/tests/graphical_isochrones_tests.py [hunk 3]
```diff
@@ -207,6 +207,15 @@ def test_graphical_isochrone_section_mode(self):
 
             assert not multi_poly_basic.contains(multi_poly_section_mode)
 
+    def test_graphical_isochrone_multi_isochrone(self):
+        q = "v1/coverage/main_routing_test/isochrones?datetime={}&from={}"
+        q = q.format('20120614T080000', r_coord)
```
Comment (azime) [inline source/jormungandr/tests/graphical_isochrones_tests.py:212]: "r_coord" is known?

File: source/jormungandr/tests/graphical_isochrones_tests.py [hunk 4]
```diff
@@ -251,7 +260,7 @@ def test_graphical_isochrones_no_duration(self):
         normal_response, error_code = self.query_no_assert(q)
 
         assert error_code == 400
-        assert normal_response['message'] == 'you should provide a \'max_duration\' argument'
+        assert normal_response['message'] == 'you should provide a \'duration\' or a \'max_duration\' argument'
```
Comment (xlqian) [inline source/jormungandr/tests/graphical_isochrones_tests.py]: minor: `"you should provide a  'duration' or a 'max_duration' argument"`

File: source/kraken/worker.cpp [hunk 5]
```diff
@@ -723,10 +723,13 @@ pbnavitia::Response Worker::graphical_isochrone(const pbnavitia::GraphicalIsochr
     }
     type::EntryPoint ep = arg.origins.empty() ? arg.destinations[0] : arg.origins[0];
 
+    std::vector<int>duration;
```
Comment (xlqian) [inline source/kraken/worker.cpp]: space before duration

File: source/routing/isochrone.cpp [hunk 6]
```diff
@@ -235,25 +235,30 @@ type::MultiPolygon build_single_isochrone(RAPTOR& raptor,
...
+                           const type::GeographicalCoord& coord_origin,
+                           const std::vector<DateTime>& bound,
+                           const map_stop_point_duration& origin,
+                           const double& speed,
+                           const std::vector<int>& duration) {
+    std::vector<Isochrone> isochrone;
+    type::MultiPolygon max_isochrone = build_single_isochrone(raptor, raptor.data.pt_data->stop_points,
+                                                              clockwise, coord_origin, bound[0], origin,
+                                                              speed, duration[0]);
+    for (unsigned int i = 1; i < duration.size(); i++) {
```
Comment (xlqian) [inline source/routing/isochrone.cpp]: size_t instead of unsigned int

File: source/routing/isochrone.cpp [hunk 7]
```diff
@@ -235,25 +235,30 @@ type::MultiPolygon build_single_isochrone(RAPTOR& raptor,
...
+                           const bool clockwise,
+                           const type::GeographicalCoord& coord_origin,
+                           const std::vector<DateTime>& bound,
+                           const map_stop_point_duration& origin,
+                           const double& speed,
+                           const std::vector<int>& duration) {
+    std::vector<Isochrone> isochrone;
+    type::MultiPolygon max_isochrone = build_single_isochrone(raptor, raptor.data.pt_data->stop_points,
+                                                              clockwise, coord_origin, bound[0], origin,
+                                                              speed, duration[0]);
```
Comment (xlqian) [inline source/routing/isochrone.cpp]: i don't know if it'd be better that you test `duration` and `bound` are not empty?

File: source/routing/isochrone.cpp [hunk 8]
```diff
@@ -235,25 +235,30 @@ type::MultiPolygon build_single_isochrone(RAPTOR& raptor,
...
+            type::MultiPolygon min_isochrone = build_single_isochrone(raptor, raptor.data.pt_data->stop_points,
+                                                                      clockwise, coord_origin, bound[i], origin,
+                                                                      speed, duration[i]);
+            boost::geometry::difference(max_isochrone, min_isochrone, output);
+            max_isochrone = std::move(min_isochrone);
+        } else {
+            output = max_isochrone;
+        }
+        Isochrone new_isochrone = Isochrone(output, duration[i], duration[i-1]);
+        isochrone.push_back(new_isochrone);
```
Comment (xlqian) [inline source/routing/isochrone.cpp]: minor: use `emplace_back` or `push_back(right value)`

File: source/jormungandr/tests/graphical_isochrones_tests.py (hunk 3 above)
Comment (ZoeBrunet) [inline source/jormungandr/tests/graphical_isochrones_tests.py:212]: Yes, it is in check_utils.py.

File: source/jormungandr/jormungandr/interfaces/v1/GraphicalIsochrone.py [hunk 9]
```diff
@@ -86,8 +87,8 @@ def get(self, region=None, uri=None):
 
         if not (args['destination'] or args['origin']):
             abort(400, message="you should provide a 'from' or a 'to' argument")
-        if not args['max_duration']:
-            abort(400, message="you should provide a 'max_duration' argument")
+        if not args['max_duration'] and not args["duration[]"]:
+            abort(400, message="you should provide a 'duration' or a 'max_duration' argument")
```
Comment (TeXitoi) [inline source/jormungandr/jormungandr/interfaces/v1/GraphicalIsochrone.py]: 'duration[]'

File: source/jormungandr/jormungandr/scenarios/simple.py [hunk 10]
```diff
@@ -228,19 +227,32 @@ def graphical_isochrones(self, request, instance):
...
         sn_params.bss_speed = request["bss_speed"]
 
-        journey_req.max_duration = request["max_duration"]
         journey_req.max_transfers = request["max_transfers"]
         self.origin_modes = request["origin_mode"]
         self.destination_modes = request["destination_mode"]
         if "forbidden_uris[]" in request and request["forbidden_uris[]"]:
             for forbidden_uri in request["forbidden_uris[]"]:
                 req.journeys.forbidden_uris.append(forbidden_uri)
+        req.isochrone.min_duration = request["min_duration"]
```
Comment (TeXitoi) [inline source/jormungandr/jormungandr/scenarios/simple.py]: not in sync with navitia-proto

File: source/jormungandr/jormungandr/scenarios/simple.py [hunk 11]
```diff
@@ -228,19 +227,32 @@ def graphical_isochrones(self, request, instance):
...
+        if request.get("max_duration"):
+            journey_req.max_duration = request["max_duration"]
+        else:
+            journey_req.max_duration = max(request["duration[]"], key=int)
+        if request.get("duration[]"):
+            for duration in sorted(request["duration[]"], key=int, reverse=True):
+                if len(request["duration[]"]) > 10:
```
Comment (TeXitoi) [inline source/jormungandr/jormungandr/scenarios/simple.py]: this test should be out of the for loop

File: source/jormungandr/jormungandr/scenarios/simple.py [hunk 12]
```diff
@@ -228,19 +227,32 @@ def graphical_isochrones(self, request, instance):
...
+                        req.isochrone.duration.append(duration)
+        req.isochrone.duration.insert(0, journey_req.max_duration)
+        req.isochrone.duration.append(req.isochrone.min_duration)
+
         journey_req.streetnetwork_params.origin_mode = self.origin_modes[0]
         journey_req.streetnetwork_params.destination_mode = self.destination_modes[0]
         resp = instance.send_and_receive(req)
         return resp
 
-
```
Comment (TeXitoi) [inline source/jormungandr/jormungandr/scenarios/simple.py:243]: This line should not be removed.

File: source/jormungandr/tests/graphical_isochrones_tests.py [hunk 13]
```diff
@@ -207,6 +207,15 @@ def test_graphical_isochrone_section_mode(self):
...
             assert not multi_poly_basic.contains(multi_poly_section_mode)
 
+    def test_graphical_isochrone_multi_isochrone(self):
+        q = "v1/coverage/main_routing_test/isochrones?datetime={}&from={}"
+        q = q.format('20120614T080000', r_coord)
+        for i in range(8):
+            q += "&duration[]={}"
+            q = q.format(i * 60)
+        response = self.query(q)
+        is_valid_graphical_isochrone(response, self.tester, q)
```
Comment (TeXitoi) [inline source/jormungandr/tests/graphical_isochrones_tests.py]:
```python
assert len(response.isochrones) == 8
```

File: source/kraken/worker.cpp [hunk 14]
```diff
@@ -723,10 +723,13 @@ pbnavitia::Response Worker::graphical_isochrone(const pbnavitia::GraphicalIsochr
     }
     type::EntryPoint ep = arg.origins.empty() ? arg.destinations[0] : arg.origins[0];
 
+    std::vector<int> duration;
```
Comment (TeXitoi) [inline source/kraken/worker.cpp]: `vector<DateTime>` seems more appropriate.

File: source/navitia-proto [hunk 15]
```diff
@@ -1 +1 @@
-Subproject commit fc2bbe33c2b67074eedecf441d8acc496748e01c
+Subproject commit 4980e20761bbafae7b49f38e37b640556e7fa0ff
```
Comment (TeXitoi) [inline source/navitia-proto]: to be updated after PR merge.

File: source/routing/isochrone.cpp [hunk 16]
```diff
@@ -235,25 +235,31 @@ type::MultiPolygon build_single_isochrone(RAPTOR& raptor,
...
+            if (duration[i] > 0) {
+                type::MultiPolygon min_isochrone = build_single_isochrone(raptor, raptor.data.pt_data->stop_points,
+                                                                          clockwise, coord_origin, bound[i], origin,
+                                                                          speed, duration[i]);
+                boost::geometry::difference(max_isochrone, min_isochrone, output);
+                max_isochrone = std::move(min_isochrone);
+            } else {
+                output = max_isochrone;
+            }
+            isochrone.push_back(Isochrone(output, duration[i], duration[i-1]));
```
Comment (TeXitoi) [inline source/routing/isochrone.cpp]:
``` c++
isochrone.push_back(Isochrone(std::move(output), duration[i], duration[i-1]));
```

File: source/routing/isochrone.h [hunk 17]
```diff
@@ -69,14 +69,25 @@ type::MultiPolygon build_single_isochrone(RAPTOR& raptor,
...
-                                    const DateTime& bound_min,
-                                    const map_stop_point_duration& origin,
-                                    const double& speed,
-                                    const int& max_duration,
-                                    const int& min_duration);
+struct Isochrone {
+    const type::MultiPolygon shape;
+    const int min_duration;
+    const int max_duration;
+    Isochrone(const type::MultiPolygon& shape,
```
Comment (TeXitoi) [inline source/routing/isochrone.h]:
``` c++
Isochrone(type::MultiPolygon shape,
```

File: source/routing/isochrone.h [hunk 18]
```diff
@@ -69,14 +69,25 @@ type::MultiPolygon build_single_isochrone(RAPTOR& raptor,
...
+              const int& min_duration,
+              const int& max_duration): shape(shape),
```
Comment (TeXitoi) [inline source/routing/isochrone.h]:
``` c++
const int& max_duration): shape(std::move(shape)),
```

File: source/routing/raptor_api.cpp [hunk 19]
```diff
@@ -1228,14 +1227,19 @@ pbnavitia::Response make_graphical_isochrone(RAPTOR &raptor,
     int day = (datetime.date() - raptor.data.meta->production_date.begin()).days();
     int time = datetime.time_of_day().total_seconds();
     DateTime init_dt = DateTimeUtils::set(day, time);
-    DateTime bound_max = clockwise ? init_dt + max_duration : init_dt - max_duration;
-    DateTime bound_min = clockwise ? init_dt + min_duration : init_dt - min_duration;
-    raptor.isochrone(departures, init_dt, bound_max, max_transfers,
+    std::vector<DateTime> bound;
+    for (int dur: duration) {
+        bound.push_back(clockwise ? init_dt + dur : init_dt - dur);
+    }
```
Comment (TeXitoi) [inline source/routing/raptor_api.cpp]: Can't that's be just calculated at the last moment to limit argument mixing?

File: source/routing/tests/isochrone_test.cpp [hunk 20]
```diff
@@ -186,29 +186,44 @@ BOOST_AUTO_TEST_CASE(build_ischrons_test) {
...
+    date_8h_9h.push_back(navitia::DateTimeUtils::set(0, "08:00"_t));
+    std::vector<navitia::DateTime> date_8h_8h30;
+    date_8h_8h30.push_back(navitia::DateTimeUtils::set(0, "08:30"_t));
+    date_8h_8h30.push_back(navitia::DateTimeUtils::set(0, "08:00"_t));
+    std::vector<navitia::DateTime> date_8h30_9h;
+    date_8h30_9h.push_back(navitia::DateTimeUtils::set(0, "09:00"_t));
+    date_8h30_9h.push_back(navitia::DateTimeUtils::set(0, "08:30"_t));
+    std::vector<int> duration_1h;
+    duration_1h.push_back(3600);
+    duration_1h.push_back(0);
```
Comment (TeXitoi) [inline source/routing/tests/isochrone_test.cpp]:
``` c++
std::vector<int> duration_1h {3600, 0};
```

File: source/routing/tests/isochrone_test.cpp [hunk 21]
```diff
@@ -186,29 +186,44 @@ BOOST_AUTO_TEST_CASE(build_ischrons_test) {
...
+    BOOST_CHECK(boost::geometry::within(coord_Notre_Dame, isochrone_9h[0].shape));
+    BOOST_CHECK(boost::geometry::within(coord_Concorde, isochrone_9h[0].shape));
+    BOOST_CHECK(boost::geometry::within(coord_Pantheon, isochrone_9h[0].shape));
+    BOOST_CHECK(boost::geometry::within(coord_Luxembourg, isochrone_9h[0].shape));
+    BOOST_CHECK(!boost::geometry::within(coord_Rennes,isochrone_9h[0].shape));
+    BOOST_CHECK(!boost::geometry::within(isochrone_8h30_9h[0].shape, isochrone_8h30[0].shape));
+    BOOST_CHECK(!boost::geometry::within(isochrone_8h30[0].shape, isochrone_8h30_9h[0].shape));
+    BOOST_CHECK(!boost::geometry::within(isochrone_9h[0].shape, isochrone_8h30_9h[0].shape));
+    BOOST_CHECK(boost::geometry::within(isochrone_8h30_9h[0].shape, isochrone_9h[0].shape));;
+    BOOST_CHECK(boost::geometry::within(circle(coord_Luxembourg, 8 * 60 * speed - 1), isochrone_9h[0].shape));
```
Comment (TeXitoi) [inline source/routing/tests/isochrone_test.cpp:225]: lack a test with several isochrones in a request.

File: source/jormungandr/jormungandr/scenarios/simple.py [hunk 22]
```diff
@@ -233,26 +233,25 @@ def graphical_isochrones(self, request, instance):
...
-                if len(request["duration[]"]) > 10:
-                    abort(400, message="you cannot provide more than 10 'duration[]'")
-                else:
-                    if req.isochrone.min_duration < duration < journey_req.max_duration:
-                        req.isochrone.duration.append(duration)
+                if request["min_duration"] < duration < journey_req.max_duration:
+                    req.isochrone.duration.append(duration)
         req.isochrone.duration.insert(0, journey_req.max_duration)
-        req.isochrone.duration.append(req.isochrone.min_duration)
+        req.isochrone.duration.append(request["min_duration"])
```
Comment (TeXitoi) [inline source/jormungandr/jormungandr/scenarios/simple.py]: what if min_duration is not setted? there is a default?

File: source/routing/isochrone.cpp [hunk 23]
```diff
@@ -241,15 +241,15 @@ std::vector<Isochrone> build_isochrones(RAPTOR& raptor,
...
+                           const std::vector<DateTime>& duration) {
     std::vector<Isochrone> isochrone;
     if (!duration.empty() && !bound.empty()) {
         type::MultiPolygon max_isochrone = build_single_isochrone(raptor, raptor.data.pt_data->stop_points,
                                                                   clockwise, coord_origin, bound[0], origin,
                                                                   speed, duration[0]);
         for (size_t i = 1; i < duration.size(); i++) {
             type::MultiPolygon output;
-            if (duration[i] > 0) {
+            if ((int)duration[i] > 0) {
```
Comment (TeXitoi) [inline source/routing/isochrone.cpp:261]: this cast should not be needed.

Comment (kinnou02) [inline source/routing/isochrone.cpp:261]: And if needed it's an old style cast, it will trigger a warning!

File: source/routing/isochrone.cpp [hunk 24]
```diff
@@ -258,7 +258,7 @@ std::vector<Isochrone> build_isochrones(RAPTOR& raptor,
             } else {
                 output = max_isochrone;
             }
-            isochrone.push_back(Isochrone(output, duration[i], duration[i-1]));
+            isochrone.insert(isochrone.begin(), Isochrone(std::move(output), duration[i], duration[i-1]));
```
Comment (TeXitoi) [inline source/routing/isochrone.cpp]: Inserting at the beginning of a vector in a loop is a bad habit. inserting at the beginning of a vector in O(n), doing that n times in a loop make it O(n²). Better to push_back() and then reverse the vector.

File: source/jormungandr/jormungandr/interfaces/v1/GraphicalIsochrone.py [hunk 25]
```diff
@@ -75,6 +75,7 @@ def __init__(self):
         super(GraphicalIsochrone, self).__init__()
         parser_get = self.parsers["get"]
         parser_get.add_argument("min_duration", type=unsigned_integer, default=0)
```
Comment (TeXitoi) [inline source/jormungandr/jormungandr/interfaces/v1/GraphicalIsochrone.py]: OK, there is a default.

File: source/jormungandr/tests/graphical_isochrones_tests.py [hunk 26]
```diff
@@ -291,3 +302,14 @@ def test_grapical_isochrone_with_from_and_to(self):
 
         assert error_code == 400
         assert normal_response['message'] == 'you cannot provide a \'from\' and a \'to\' argument'
+
+    def test_grapical_isochrone_more_than_10_duration(self):
+        q = "v1/coverage/main_routing_test/isochrones?datetime={}&from={}"
+        q = q.format('20120614T080000', r_coord)
+        for i in range(20):
+            q += "&duration[]={}"
+            q = q.format(i*60)
```
Comment (TeXitoi) [inline source/jormungandr/tests/graphical_isochrones_tests.py]: If there is `{}` in q before, that's buggy.

better to 

``` python
q += "&duration[]={}".format(i*60)
```

File: source/navitia-proto [hunk 27]
```diff
@@ -1 +1 @@
-Subproject commit fc2bbe33c2b67074eedecf441d8acc496748e01c
+Subproject commit 7aead832c330f4df5dfbf08a07933d6126bb5ea8
```
Comment (TeXitoi) [inline source/navitia-proto]: To be updated when navitia-proto is merged.

File: source/routing/isochrone.h [hunk 28]
```diff
@@ -69,14 +69,27 @@ type::MultiPolygon build_single_isochrone(RAPTOR& raptor,
...
-                                    const double& speed,
-                                    const int& max_duration,
-                                    const int& min_duration);
+struct Isochrone {
+    type::MultiPolygon shape;
+    DateTime min_duration;
+    DateTime max_duration;
+    Isochrone(type::MultiPolygon shape,
+              DateTime min_duration,
+              DateTime max_duration): shape(shape),
```
Comment (TeXitoi) [inline source/routing/isochrone.h]:
``` c++
               DateTime max_duration): shape(std::move(shape)), 
```

File: source/routing/isochrone.h [hunk 29]
```diff
@@ -69,14 +69,27 @@ type::MultiPolygon build_single_isochrone(RAPTOR& raptor,
...
+                                      min_duration(min_duration),
+                                      max_duration(max_duration){
+
+    }
+};
+
+
+
```
Comment (TeXitoi) [inline source/routing/isochrone.h]: 3 empty lines is a bit too much.

File: source/routing/raptor_api.cpp [hunk 30]
```diff
@@ -1228,14 +1227,19 @@ pbnavitia::Response make_graphical_isochrone(RAPTOR &raptor,
     int day = (datetime.date() - raptor.data.meta->production_date.begin()).days();
     int time = datetime.time_of_day().total_seconds();
     DateTime init_dt = DateTimeUtils::set(day, time);
-    DateTime bound_max = clockwise ? init_dt + max_duration : init_dt - max_duration;
-    DateTime bound_min = clockwise ? init_dt + min_duration : init_dt - min_duration;
-    raptor.isochrone(departures, init_dt, bound_max, max_transfers,
+    std::vector<DateTime> bound;
+    for (int dur: duration) {
```
Comment (TeXitoi) [inline source/routing/raptor_api.cpp]: duration is a vector of DateTime, not int. better to use auto.

**Comment by TeXitoi** [comment]
OK for me when navitia-proto is merged.

**Pull request merged**